src/test/ledger/View_test.cpp

```diff
@@ -384,6 +384,279 @@ class View_test : public beast::unit_test::suite
+                ledger->stateMap().lower_bound(uint256(40))->key() ==
+                uint256(39));
+            BEAST_EXPECT(
+                ledger->stateMap().lower_bound(uint256(47))->key() ==
+                uint256(46));
+            BEAST_EXPECT(
+                ledger->stateMap().lower_bound(uint256(48))->key() ==
+                uint256(47));
+            BEAST_EXPECT(
+                ledger->stateMap().lower_bound(uint256(64))->key() ==
+                uint256(48));
+
+            BEAST_EXPECT(
+                ledger->stateMap().lower_bound(uint256(90))->key() ==
+                uint256(66));
+            BEAST_EXPECT(
+                ledger->stateMap().lower_bound(uint256(96))->key() ==
+                uint256(66));
+            BEAST_EXPECT(
+                ledger->stateMap().lower_bound(uint256(100))->key() ==
+                uint256(66));
+
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(0))->key() ==
+                uint256(1));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(5))->key() ==
+                uint256(6));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(15))->key() ==
+                uint256(16));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(16))->key() ==
+                uint256(20));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(18))->key() ==
+                uint256(20));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(20))->key() ==
+                uint256(25));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(31))->key() ==
+                uint256(32));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(32))->key() ==
+                uint256(33));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(47))->key() ==
+                uint256(48));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(48))->key() ==
+                uint256(66));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(53))->key() ==
+                uint256(66));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(66))->key() ==
+                uint256(100));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(70))->key() ==
+                uint256(100));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(85))->key() ==
+                uint256(100));
+            BEAST_EXPECT(
+                ledger->stateMap().upper_bound(uint256(98))->key() ==
+                uint256(100));
+            BEAST_EXPECT(ledger->stateMap().upper_bound(uint256(100)) == e);
+            BEAST_EXPECT(ledger->stateMap().upper_bound(uint256(155)) == e);
+        }
+    }
```

Comment:
I appreciate that there are tests, but I'd like to try to be a little more judicious with the code (less to read, less to compile, less to run). The tests here seem to span a few categories:

- When the ID is in vs not in the SHAMap.
- When the request is for lower vs upper bound.
- When the requested ID is in vs not in the SHAMap.
- When the requested ID has the same parent inner node vs sibling parent inner node vs distant parent inner node.

So that's 24 tests that can be satisfied with one example SHAMap. Would be nice to put the test strategy in a comment, if not going so far as to name them separately.

Reply:
Not all of the assertions in the last test are necessary either, so that's not the solution I'm asking for. My guiding principle is this: what does this assertion tell us that no other assertion tells us? If an assertion provides no new information, then it is a duplicate and should be eliminated. Writing a test that is just like another test but bigger is not helpful. It adds more code to read, compile, run, and maintain without adding any value.

Here's what I would write, including the comment:

```cpp
// The tests here try to span a few categories:
// - When the given ID is in vs not in the SHAMap.
// - When the request is for lower vs upper bound.
// - When the requested ID is in vs not in the SHAMap.
// - When the given ID and requested ID are have the same parent inner node vs sibling parent inner nodes vs distant parent inner nodes.

setup({
    0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15,
    16,
    32,
    256});
auto const& state = ledger->stateMap();
auto const& end = state.end();

BEAST_EXPECT(state.lower_bound(uint256(0)) == end);
BEAST_EXPECT(state.lower_bound(uint256(1)) == uint256(0));
BEAST_EXPECT(state.lower_bound(uint256(16)) == uint256(15));
BEAST_EXPECT(state.lower_bound(uint256(32)) == uint256(16));
BEAST_EXPECT(state.lower_bound(uint256(64)) == uint256(32));
BEAST_EXPECT(state.lower_bound(uint256(512)) == uint256(256));

BEAST_EXPECT(state.upper_bound(uint256(512)) == end);
BEAST_EXPECT(state.upper_bound(uint256(256)) == end);
BEAST_EXPECT(state.upper_bound(uint256(64)) == uint256(256));
BEAST_EXPECT(state.upper_bound(uint256(32)) == uint256(256));
BEAST_EXPECT(state.upper_bound(uint256(16)) == uint256(32));
BEAST_EXPECT(state.upper_bound(uint256(15)) == uint256(16));
BEAST_EXPECT(state.upper_bound(uint256(0)) == uint256(1));
```

Reply:
I would rather have too many tests than too little. If there is a case that we are missing, let me know, but I'm not really comfortable removing these tests that appear unnecessary. The SHAMap is not a trivial data structure. A bigger tree with more branches is different than a smaller tree with less, and I would prefer to test both cases, even if it seems like that's not necessary.